src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,152 @@
+                Moniker = GetMoniker(),
+                MaxHeight = lineHeight,
+                Margin = new Thickness(1, 0, 5, 0)
+            };
+            var stackPanel = new StackPanel
+            {
+                Height = lineHeight,
+                Orientation = Orientation.Horizontal
+            };
+
+            stackPanel.Children.Add(image);
+            stackPanel.Children.Add(block);
+
+            var border = new Border
+            {
+                BorderBrush = format.BackgroundBrush,
+                BorderThickness = new Thickness(1),
+                Background = Brushes.Transparent,
+                Child = stackPanel,
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(10, top: 0, 0, bottom: 1),
+                Padding = new Thickness(1)
+            };
+
+            var editorBackground = (Color)_editorFormatMap.GetProperties("TextView Background")["BackgroundColor"];
+            ImageThemingUtilities.SetImageBackgroundColor(border, editorBackground);
+
+            border.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            view.ViewportWidthChanged += ViewportWidthChangedHandler;
+
+            return new GraphicsResult(border, dispose:
+                () =>
+                {
+                    link.RequestNavigate -= HandleRequestNavigate;
+                    view.ViewportWidthChanged -= ViewportWidthChangedHandler;
+                });
+
+            void ViewportWidthChangedHandler(object s, EventArgs e)
+            {
+                if (Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                {
+                    Canvas.SetLeft(border, view.ViewportWidth - border.DesiredSize.Width);
+                }
+            }
+
+            void HandleRequestNavigate(object sender, RoutedEventArgs e)
+            {
+                var uri = link.NavigateUri;
+                _ = _navigateToLinkService.TryNavigateToLinkAsync(uri, CancellationToken.None);
+                e.Handled = true;
+            }
+        }
+
+        private ImageMoniker GetMoniker()
+        {
+            switch (_diagnostic.Severity)
+            {
+                case DiagnosticSeverity.Warning:
+                    return KnownMonikers.StatusWarning;
+                default:
+                    return KnownMonikers.StatusError;
+            }
+        }
+
+        /// <summary>
+        /// Gets called when the ClassificationFormatMap is changed to update the adornment
+        /// </summary>
+        public void UpdateColor(TextFormattingRunProperties format, UIElement adornment)
+        {
+            var border = (Border)adornment;
+            border.BorderBrush = format.BackgroundBrush;
+            var block = (TextBlock)border.Child;
+            block.Foreground = format.ForegroundBrush;
```

Comment:
i don't see this accessing instance state.  Does this need to be an instance method?

Reply:
yes, what is the best way to deal with the warning here? is it fine to suppress it?

Reply:
?

not sure what you mean.  what warning?

Reply:
It gives a warning and says that it can be made static, but it gets called at the InlineDiagnosticsAdornmentManager. So if I make it static there is going to be an error.